File: src/main/java/com/prototype/framework/configuration/mvc/converter/MvcSingleValueObjectConverter.java
package com.prototype.framework.configuration.mvc.converter;

import com.prototype.architecture.layer_03_domain.SingleValueObject;
import lombok.SneakyThrows;
import org.springframework.core.convert.TypeDescriptor;
import org.springframework.core.convert.converter.GenericConverter;
import org.springframework.stereotype.Component;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestParam;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * ValueObjectを継承したクラスを
 * {@link RequestParam}や{@link PathVariable}で使用するためのコンバーター
 */
@Component
public class MvcSingleValueObjectConverter implements GenericConverter {

    @Override
    public Set<ConvertiblePair> getConvertibleTypes() {
        final ConvertiblePair[] pairs = new ConvertiblePair[] {
                new ConvertiblePair(String.class, SingleValueObject.class),
                new ConvertiblePair(Number.class, SingleValueObject.class),
        };
        return new HashSet(List.of(pairs));
    }

    @SneakyThrows
    @Override
    public Object convert(Object source, TypeDescriptor sourceType, TypeDescriptor targetType) {
        if (source instanceof String o && targetType.getType().getSuperclass() == SingleValueObject.class) {
            final Type genericType = ((ParameterizedType) targetType.getType().getGenericSuperclass()).getActualTypeArguments()[0];
            if (genericType == String.class) {
                return targetType.getType().getDeclaredConstructor(String.class).newInstance(o);
            }
            if (genericType == Integer.class) {
                return targetType.getType().getDeclaredConstructor(Integer.class).newInstance(Integer.parseInt(o));
            }
        }

        return source;
    }

}

File: src/main/kotlin/com/prototype/framework/configuration/mvc/converter/MvcSingleValueObjectConverter.kt
package com.prototype.framework.configuration.mvc.converter;

import com.prototype.architecture.layer_03_domain.SingleValueObject
import com.prototype.framework.extension.toSingleValueObject
import org.springframework.core.convert.TypeDescriptor
import org.springframework.core.convert.converter.GenericConverter
import org.springframework.core.convert.converter.GenericConverter.ConvertiblePair
import org.springframework.stereotype.Component
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * ValueObjectを継承したクラスを
 * {@link RequestParam}や{@link PathVariable}で使用するためのコンバーター
 */
@Component
class MvcSingleValueObjectConverter : GenericConverter {

    override fun getConvertibleTypes(): MutableSet<ConvertiblePair>? {
        return listOf(
                ConvertiblePair(String::class.java, SingleValueObject::class.java),
                ConvertiblePair(Number::class.java, SingleValueObject::class.java)
        ).toMutableSet()
    }

    override fun convert(source: Any?, sourceType: TypeDescriptor, targetType: TypeDescriptor): Any? {
        return source?.toSingleValueObject(sourceType, targetType)
    }

}